tombstone_gc.cc

```diff
@@ -0,0 +1,164 @@
+        dblog.trace("Get gc_before for ks={}, table={}, range={}, mode=timeout", s->ks_name(), s->cf_name(), range);
+        auto gc_before = saturating_subtract(query_time, s->gc_grace_seconds());
+        return {gc_before, gc_before, knows_entire_range};
+    }
+    case tombstone_gc_mode::disabled: {
+        dblog.trace("Get gc_before for ks={}, table={}, range={}, mode=disabled", s->ks_name(), s->cf_name(), range);
+        return {gc_clock::time_point::min(), gc_clock::time_point::min(), knows_entire_range};
+    }
+    case tombstone_gc_mode::immediate: {
+        dblog.trace("Get gc_before for ks={}, table={}, range={}, mode=immediate", s->ks_name(), s->cf_name(), range);
+        return {gc_clock::time_point::max(), gc_clock::time_point::max(), knows_entire_range};
+    }
+    case tombstone_gc_mode::repair: {
+        const std::chrono::seconds& propagation_delay = options.propagation_delay_in_seconds();
+        auto min_gc_before = gc_clock::time_point::min();
+        auto max_gc_before = gc_clock::time_point::min();
+        auto min_repair_timestamp = gc_clock::time_point::min();
+        auto max_repair_timestamp = gc_clock::time_point::min();
+        int hits = 0;
+        knows_entire_range = false;
+        auto m = get_repair_history_map_for_table(s->id());
+        if (m) {
+            auto interval = locator::token_metadata::range_to_interval(range);
+            auto min = gc_clock::time_point::max();
+            auto max = gc_clock::time_point::min();
+            bool contains_all = false;
+            for (auto& x : boost::make_iterator_range(m->map.equal_range(interval))) {
+                auto r = locator::token_metadata::interval_to_range(x.first);
+                min = std::min(x.second, min);
+                max = std::max(x.second, max);
+                if (++hits == 1 && r.contains(range, dht::tri_compare)) {
+                    contains_all = true;
+                }
+            }
+            if (hits == 0) {
+                min_repair_timestamp = gc_clock::time_point::min();
+                max_repair_timestamp = gc_clock::time_point::min();
+            } else {
+                knows_entire_range = hits == 1 && contains_all;
+                min_repair_timestamp = min;
+                max_repair_timestamp = max;
+            }
+            min_gc_before = saturating_subtract(min_repair_timestamp, propagation_delay);
+            max_gc_before = saturating_subtract(max_repair_timestamp, propagation_delay);
+        };
+        dblog.info("Get gc_before for ks={}, table={}, range={}, mode=repair, min_repair_timestamp={}, max_repair_timestamp={}, propagation_delay={}, min_gc_before={}, max_gc_before={}, hits={}, knows_entire_range={}",
+                s->ks_name(), s->cf_name(), range, min_repair_timestamp, max_repair_timestamp, propagation_delay.count(), min_gc_before, max_gc_before, hits, knows_entire_range);
+        return {min_gc_before, max_gc_before, knows_entire_range};
```

Comment:
What's the meaning of "knows entire range"? Strictly speaking, repair history always knows all ranges.

Reply:
It means the range given by sstable is a subset of a single range in the repair history map. Note, this is used only by sstable::get_gc_before_for_fully_expire.